src/vendor/vendor.c

```diff
@@ -1441,74 +1474,41 @@ int RDK_GetGroup(int group_id, kv_vector_t *params)
 int RDK_SetGroup(int group_id, kv_vector_t *params, unsigned *param_types, int *failure_index)
 {
     int i;
-    int ccsp_err;
-    rdk_component_t *rdkc;
-    parameterValStruct_t *rdk_params;
-    parameterValStruct_t *pvs;
+    int rbus_err;
     int err = USP_ERR_OK;
-    char *fault_param = NULL;
+    rbusValue_t rbus_val;
     kv_pair_t *kv;
-    CCSP_MESSAGE_BUS_INFO *bus_info = (CCSP_MESSAGE_BUS_INFO *)bus_handle;
+    char *s = NULL;
 
     // Exit if this function is called before D-Bus has been connected to
     if (bus_handle == NULL)
     {
-        USP_ERR_SetMessage("%s: called before connected to D-Bus", __FUNCTION__);
+        USP_ERR_SetMessage("%s: called before connected to R-Bus", __FUNCTION__);
         return USP_ERR_INTERNAL_ERROR;
     }
 
     // Form the rdk_params structure to pass to CcspBaseIf_setParameterValues()
-    rdkc = &rdk_components[group_id];
-    rdk_params = USP_MALLOC(params->num_entries * sizeof(parameterValStruct_t));
     for (i=0; i < params->num_entries; i++)
     {
-        pvs = &rdk_params[i];
         kv = &params->vector[i];
-        pvs->parameterName = kv->key;
-        pvs->parameterValue = kv->value;
-        pvs->type = UspTypeToRdkType(param_types[i]);
-    }
+	rbusValue_Init(&rbus_val);
+	s=kv->key;
+	rbusValue_SetString(rbus_val, kv->value);
+	rbus_err = rbus_set(bus_handle,s,rbus_val, NULL);
```

Comment:
Addressed issue with a session using rbus_set()
Please review.

Reply:
The loop should exit at the first parameter which fails to set, and it must set *failure_index to indicate that parameter.

But you aren't really using a single session for all parameters here, because you're committing the session after setting each parameter (isCommit=true)
The required behavior is that if any one parameter fails to set, then none of the parameters should be set in the list ie if you've already set them successfully, then they should be rolled back to their previous values. Basically the set operation is supposed to be atomic - either all are set or none are set, upon return of the function. Can this be achieved using isCommit=false, then calling some other rbus function after successfully setting all parameters in the list to perform the final commit ? What about rollback though ? R-Bus sessions don't seem to support atomicity because I can't see any function which rolls back the current session.
I assume that rbus_setMulti() is atomic. However it doesn't tell you which parameter failed to set.
Please could you look through the R-Bus API and speak to the R-Bus architects to see if there's a better solution to this. As it stands this functionality seems worse than under CCSP API (which I believe was atomic and which also told you the first parameter which failed to set).
By the way I don't think that the solution is to read the values before setting them, just in case we need to roll them back - that would be very inefficient and might not work anyway (because the rollback SPV could fail).